File: src/clase7/src/hello_moveit_obstacles.cpp
#include <geometry_msgs/msg/pose_stamped.hpp>
#include <memory>
#include <string>
#include <vector>
#include <rclcpp/rclcpp.hpp>
#include <moveit/move_group_interface/move_group_interface.hpp>
#include <moveit/planning_scene_interface/planning_scene_interface.hpp>
#include <moveit_msgs/msg/collision_object.hpp>
#include <shape_msgs/msg/solid_primitive.hpp>

// =====================================================================
// GOAL - modificar estos valores
// =====================================================================
static constexpr char PLANNING_GROUP[] = "arm";
static constexpr char EE_LINK[] = "tool0";
static constexpr char REF_FRAME[] = "base_link";

static constexpr double GOAL_X  = 0.261;
static constexpr double GOAL_Y  = -0.176;
static constexpr double GOAL_Z  = 0.168;
static constexpr double GOAL_QX = 1.0;
static constexpr double GOAL_QY = 0.0;
static constexpr double GOAL_QZ = 0.0;
static constexpr double GOAL_QW = 0.0;
// =====================================================================

// =====================================================================
// OBSTACLES - modificar / agregar / borrar entradas libremente.
//   type: 'b' = box (dims = {sx, sy, sz})
//         's' = sphere (dims = {radius})
//   pose está en el frame REF_FRAME, orientación identidad.
// =====================================================================
struct Obstacle
{
  std::string id;
  char type;
  double x, y, z;
  std::vector<double> dims;
};

static const std::vector<Obstacle> OBSTACLES = {
  {"mesa",   'b', 0.20,  0.00, 0.02, {0.30, 0.30, 0.02}},
  {"pilar",  'b', 0.15, -0.10, 0.15, {0.03, 0.03, 0.20}},
  {"pelota", 's', 0.05, -0.15, 0.20, {0.04}},
};
// =====================================================================

static moveit_msgs::msg::CollisionObject make_collision_object(
  const Obstacle & o, const std::string & frame_id)
{
  moveit_msgs::msg::CollisionObject co;
  co.header.frame_id = frame_id;
  co.id = o.id;

  shape_msgs::msg::SolidPrimitive prim;
  if (o.type == 's') {
    prim.type = shape_msgs::msg::SolidPrimitive::SPHERE;
    prim.dimensions = {o.dims.at(0)};
  } else {
    prim.type = shape_msgs::msg::SolidPrimitive::BOX;
    prim.dimensions = {o.dims.at(0), o.dims.at(1), o.dims.at(2)};
  }

  geometry_msgs::msg::Pose pose;
  pose.position.x = o.x;
  pose.position.y = o.y;
  pose.position.z = o.z;
  pose.orientation.w = 1.0;

  co.primitives.push_back(prim);
  co.primitive_poses.push_back(pose);
  co.operation = moveit_msgs::msg::CollisionObject::ADD;
  return co;
}

int main(int argc, char * argv[])
{
  rclcpp::init(argc, argv);

  auto const node = std::make_shared<rclcpp::Node>(
    "hello_moveit_obstacles",
    rclcpp::NodeOptions().automatically_declare_parameters_from_overrides(true)
  );

  auto const logger = rclcpp::get_logger("hello_moveit_obstacles");

  using moveit::planning_interface::MoveGroupInterface;
  using moveit::planning_interface::PlanningSceneInterface;

  auto arm_group_interface = MoveGroupInterface(node, PLANNING_GROUP);

  arm_group_interface.setPlanningPipelineId("ompl");
  arm_group_interface.setPlannerId("RRTConnectkConfigDefault");
  arm_group_interface.setPlanningTime(5.0);
  arm_group_interface.setMaxVelocityScalingFactor(1.0);
  arm_group_interface.setMaxAccelerationScalingFactor(1.0);
  arm_group_interface.setPoseReferenceFrame(REF_FRAME);
  arm_group_interface.setEndEffectorLink(EE_LINK);

  RCLCPP_INFO(logger, "Planning pipeline: %s", arm_group_interface.getPlanningPipelineId().c_str());
  RCLCPP_INFO(logger, "Planner ID: %s", arm_group_interface.getPlannerId().c_str());
  RCLCPP_INFO(logger, "Planning time: %.2f", arm_group_interface.getPlanningTime());

  PlanningSceneInterface psi;
  std::vector<moveit_msgs::msg::CollisionObject> objects;
  objects.reserve(OBSTACLES.size());
  for (auto const & o : OBSTACLES) {
    objects.push_back(make_collision_object(o, REF_FRAME));
  }
  bool const scene_ok = psi.applyCollisionObjects(objects);
  RCLCPP_INFO(logger, "Obstáculos publicados: %zu (ok=%s)",
              objects.size(), scene_ok ? "true" : "false");

  auto const target_pose = [&node]{
    geometry_msgs::msg::PoseStamped msg;
    msg.header.frame_id = REF_FRAME;
    msg.header.stamp = node->now();
    msg.pose.position.x = GOAL_X;
    msg.pose.position.y = GOAL_Y;
    msg.pose.position.z = GOAL_Z;
    msg.pose.orientation.x = GOAL_QX;
    msg.pose.orientation.y = GOAL_QY;
    msg.pose.orientation.z = GOAL_QZ;
    msg.pose.orientation.w = GOAL_QW;
    return msg;
  }();
  arm_group_interface.setPoseTarget(target_pose);

  auto const [success, plan] = [&arm_group_interface] {
    moveit::planning_interface::MoveGroupInterface::Plan msg;
    auto const ok = static_cast<bool>(arm_group_interface.plan(msg));
    return std::make_pair(ok, msg);
  }();

  if (success)
  {
    RCLCPP_INFO(logger, "Plan encontrado, ejecutando...");
    arm_group_interface.execute(plan);
  }
  else
  {
    RCLCPP_ERROR(logger, "Planning failed!");
  }

  rclcpp::shutdown();
  return 0;
}

File: src/clase7/scripts/hello_moveit_obstacles.py
#!/usr/bin/env python3
# Variante de hello_moveit.py que además de mover el myCobot 320 a una pose,
# publica un PlanningScene con obstáculos (cajas y esferas) al servicio
# /apply_planning_scene. El planner de OMPL/RRTConnect debe rodearlos para
# llegar al target.

import rclpy
from rclpy.node import Node
from rclpy.action import ActionClient
from moveit_msgs.action import MoveGroup
from moveit_msgs.msg import (MotionPlanRequest, WorkspaceParameters,
                              Constraints, PositionConstraint,
                              OrientationConstraint, BoundingVolume,
                              PlanningScene, CollisionObject)
from moveit_msgs.srv import ApplyPlanningScene
from geometry_msgs.msg import Pose, Vector3
from shape_msgs.msg import SolidPrimitive

# =====================================================================
# GOAL - modificar estos valores (deben coincidir con hello_moveit_obstacles.cpp)
# =====================================================================
PLANNING_GROUP = 'arm'
EE_LINK = 'tool0'
REF_FRAME = 'base_link'

GOAL_X = 0.061
GOAL_Y = -0.176
GOAL_Z = 0.168
GOAL_QX = 1.0
GOAL_QY = 0.0
GOAL_QZ = 0.0
GOAL_QW = 0.0
# =====================================================================

# =====================================================================
# OBSTACLES - modificar / agregar / borrar entradas libremente.
#   type: 'box'    -> dims = (sx, sy, sz)
#         'sphere' -> dims = (radius,)
#   pose está en el frame REF_FRAME, orientación identidad.
# =====================================================================
OBSTACLES = [
    {'id': 'mesa',   'type': 'box',    'pose': (0.20,  0.00, 0.02), 'dims': (0.30, 0.30, 0.02)},
    {'id': 'pilar',  'type': 'box',    'pose': (0.15, -0.10, 0.15), 'dims': (0.03, 0.03, 0.20)},
    {'id': 'pelota', 'type': 'sphere', 'pose': (0.05, -0.15, 0.20), 'dims': (0.04,)},
]
# =====================================================================


def _make_collision_object(spec):
    co = CollisionObject()
    co.header.frame_id = REF_FRAME
    co.id = spec['id']

    prim = SolidPrimitive()
    if spec['type'] == 'sphere':
        prim.type = SolidPrimitive.SPHERE
        prim.dimensions = [float(spec['dims'][0])]
    else:
        prim.type = SolidPrimitive.BOX
        prim.dimensions = [float(d) for d in spec['dims']]

    pose = Pose()
    pose.position.x = float(spec['pose'][0])
    pose.position.y = float(spec['pose'][1])
    pose.position.z = float(spec['pose'][2])
    pose.orientation.w = 1.0

    co.primitives.append(prim)
    co.primitive_poses.append(pose)
    co.operation = CollisionObject.ADD
    return co


class HelloMoveitObstacles(Node):
    def __init__(self):
        super().__init__('hello_moveit_obstacles')
        self._client = ActionClient(self, MoveGroup, '/move_action')
        self._scene_client = self.create_client(ApplyPlanningScene, '/apply_planning_scene')

    def publish_obstacles(self):
        if not self._scene_client.wait_for_service(timeout_sec=10.0):
            self.get_logger().warn(
                '/apply_planning_scene no disponible, el planner correrá sin escena')
            return

        scene = PlanningScene()
        scene.is_diff = True
        scene.world.collision_objects = [_make_collision_object(o) for o in OBSTACLES]

        req = ApplyPlanningScene.Request()
        req.scene = scene

        future = self._scene_client.call_async(req)
        rclpy.spin_until_future_complete(self, future)

        result = future.result()
        ok = bool(result and result.success)
        self.get_logger().info(
            f'Obstáculos publicados: {len(scene.world.collision_objects)} (ok={ok})')

    def send_goal(self):
        self._client.wait_for_server()
        self.get_logger().info('move_group disponible, enviando goal...')

        goal = MoveGroup.Goal()
        req = MotionPlanRequest()

        req.group_name = PLANNING_GROUP
        req.num_planning_attempts = 5
        req.allowed_planning_time = 5.0
        req.max_velocity_scaling_factor = 1.0
        req.max_acceleration_scaling_factor = 1.0
        req.planner_id = 'RRTConnect'

        req.workspace_parameters = WorkspaceParameters()
        req.workspace_parameters.header.frame_id = REF_FRAME
        req.workspace_parameters.min_corner.x = -2.0
        req.workspace_parameters.min_corner.y = -2.0
        req.workspace_parameters.min_corner.z = -2.0
        req.workspace_parameters.max_corner.x = 2.0
        req.workspace_parameters.max_corner.y = 2.0
        req.workspace_parameters.max_corner.z = 2.0

        constraints = Constraints()

        pos_constraint = PositionConstraint()
        pos_constraint.header.frame_id = REF_FRAME
        pos_constraint.link_name = EE_LINK
        pos_constraint.target_point_offset = Vector3(x=0.0, y=0.0, z=0.0)

        bv = BoundingVolume()
        sp = SolidPrimitive()
        sp.type = SolidPrimitive.SPHERE
        sp.dimensions = [0.01]  # tolerancia 1cm
        bv.primitives.append(sp)

        goal_pose = Pose()
        goal_pose.position.x = GOAL_X
        goal_pose.position.y = GOAL_Y
        goal_pose.position.z = GOAL_Z
        bv.primitive_poses.append(goal_pose)

        pos_constraint.constraint_region = bv
        pos_constraint.weight = 1.0
        constraints.position_constraints.append(pos_constraint)

        ori_constraint = OrientationConstraint()
        ori_constraint.header.frame_id = REF_FRAME
        ori_constraint.link_name = EE_LINK
        ori_constraint.orientation.x = GOAL_QX
        ori_constraint.orientation.y = GOAL_QY
        ori_constraint.orientation.z = GOAL_QZ
        ori_constraint.orientation.w = GOAL_QW
        ori_constraint.absolute_x_axis_tolerance = 0.1
        ori_constraint.absolute_y_axis_tolerance = 0.1
        ori_constraint.absolute_z_axis_tolerance = 0.1
        ori_constraint.weight = 1.0
        constraints.orientation_constraints.append(ori_constraint)

        req.goal_constraints.append(constraints)
        goal.request = req
        goal.planning_options.plan_only = False  # planifica y ejecuta

        future = self._client.send_goal_async(goal)
        rclpy.spin_until_future_complete(self, future)

        goal_handle = future.result()
        if not goal_handle.accepted:
            self.get_logger().error('Goal rechazado')
            return

        self.get_logger().info('Goal aceptado, ejecutando...')
        result_future = goal_handle.get_result_async()
        rclpy.spin_until_future_complete(self, result_future)

        result = result_future.result().result
        if result.error_code.val == 1:
            self.get_logger().info('Movimiento completado exitosamente')
        else:
            self.get_logger().error(f'Error: error_code={result.error_code.val}')


def main():
    rclpy.init()
    node = HelloMoveitObstacles()
    node.publish_obstacles()
    node.send_goal()
    rclpy.shutdown()


if __name__ == '__main__':
    main()
